feat/add bp test doc: adds the first draft of the user guide for the test framework (#983)
GoogleCloudPlatform/cloud-foundation-toolkit: infra/blueprint-test/README.md

# 1. Background and Overview

Integration testing has been instrumental in establishing a complete and functional CI pipeline. Similar to most quality software engineering modules and applications, GCP Blueprints employ an integration testing framework that is triggered as part of the CI process as well.

Apart from the necessity of including a testing framework as part of our GCP blueprints we took additional steps to ensure the following:

- Easy to learn and to setup as part of the blueprints development process
- Compatible with both Terraform and KRM based GCP blueprints

Considering the above, our test framework has been developed (details in the following sections) with backward compatibility to allow for current tests to keep functioning.

> ```

Reviewer: nit: this does not need to be a codeblock. Lets use maybe italic to highlight any non code portions?

Author: Acked.

For the purpose of this user guide, the [terraform-google-sql-db blueprint](https://github.com/terraform-google-modules/terraform-google-sql-db) is used to demonstrate the usage of the framework.

## 2.1 Blueprint directory structure

```
├── ...
├── modules/submodules/*.tf
├── examples
│   ├── mssql-public/*.tf
│   ├── mysql-ha/*.tf
│   ├── mysql-private/*.tf
│   ├── mysql-public/*.tf
│   ├── postgresql-ha/*.tf
│   ├── postgresql-public/*.tf
│   └── postgresql-public-iam/*.tf
├── test
│   ├── fixtures
│   │   ├── mssql-ha/*.tf
│   │   ├── mysql-ha/*.tf
│   │   ├── mysql-private/*.tf
│   │   ├── mysql-public/*.tf
│   │   ├── postgresql-ha/*.tf
│   │   ├── postgresql-public/*.tf
│   │   └── postgresql-public-iam/*.tf
│   ├── integration
│   │   ├── testing.yaml
│   │   ├── discover_test.go
│   │   ├── mssql-ha/*
│   │   ├── mysql-ha/*
│   │   ├── mysql-public/*
│   │   │   └── mysql_public_test.go
│   │   ├── postgresql-ha/*
│   │   ├── postgresql-public/*
│   │   ├── postgresql-public-iam/*
│   └── setup/*.tf
└── ...
```

> Before jumping into the specifics for test development, let’s review how the blueprint directory is structured.

Reviewer: ```suggestion
Let’s review how the blueprint directory is structured.
```

Author: Done.

> - `examples`  - this directory holds examples that may call the main blueprint or sub-blueprints (for Terraform, these are in the  modules  directory).

Reviewer: ```suggestion
- `examples`  - this directory holds examples that may call the main blueprint or sub-blueprints (for Terraform, this is the main module or sub-modules within the modules directory).
```

Author: Changed.

- `test/fixtures`  - this directory contains “fixture” configuration. In most cases, this should be configured to wrap examples that need additional inputs and unify the interface for testing. Usage of fixtures is discouraged unless necessary.
- `test/integration`  - this directory is intended to hold integration tests that are responsible for running and asserting test values for a given fixture.

> - `test/setup`  - this directory holds the GCP project and initial resource setup modules that are a prerequisite for the blueprint tests to run.

Reviewer: ```suggestion
- `test/setup`  - this directory holds configuration for creating the GCP project and initial resources that are a prerequisite for the blueprint tests to run.
```

Author: Chagned.

The first step in writing the test is to wire it up with the required packages and methods signatures that the test framework expects as follows:
1. Cd in the `test/integration/mysql-public` directory or create it if it's not present already in the blueprint and the cd into it.
2. Create file in `mysql_public_test.go` with the following content: 

As a good practice use this convention to name your test files: <example_name>_test.go
     
```go
// define test package name
package mysql_public
 
import (
  "fmt"
  "testing"
 
  // import the blueprints test framework modules for testing and assertions
  "github.com/GoogleCloudPlatform/cloud-foundation-toolkit/infra/blueprint-test/pkg/gcloud"   
  "github.com/GoogleCloudPlatform/cloud-foundation-toolkit/infra/blueprint-test/pkg/tft"
  "github.com/stretchr/testify/assert"
)
 
// name the function as Test*
func TestMySqlPublicModule(t *testing.T) {
   ...
   // initialize Terraform test from the blueprint test framework
   mySqlT := tft.NewTFBlueprintTest(t)
   // define and write a custom verifier for this test case call the default verify for confirming no additional changes
   mySqlT.DefineVerify(func(assert *assert.Assertions) {

>        mySqlT.DefaultVerify(assert)    

Reviewer: lets add a comment about what this does

Author: Added.

       ...   
   })
   // call the test function to execute the integration test
   mySqlT.Test()
}
```

The next step in the process is to write the logic for assertions.

1. In most cases, you will be asserting against values retrieved from the GCP environment. This can be done by using the (gcloud)[https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/blob/master/infra/blueprint-test/pkg/gcloud/gcloud.go] helper in our test framework, which executes gcloud commands and stores their JSON outputs as. The gcloud helper can be initialized as follows:

```go
// The tft struct can be used to pull output variables of the TF module being invoked by this test
op := gcloud.Run(t, fmt.Sprintf("sql instances describe %s --project %s", mySqlT.GetStringOutput("name"), mySqlT.GetStringOutput("project_id")))
```

2. Once you have retrieved values from GCP, use the (assert)[https://pkg.go.dev/github.com/stretchr/testify/assert] package to perform custom validations with respect to the resources provisioned. Here are some common assertions that can be useful in most test scenarios.

   1. Contains

    ```go
    // assert values that are contained in the expected output

>     assert.Contains(op.Get("gceZone").String(), region, "GCE region is valid")

Reviewer: for `op.Get("gceZone").String()` shouldn't it be an assert.Equals. Contains would be for a slice?

Author: It can assert part of a string, list or map. I just picked gceZone as an example since it would always contain the regional sub-string.

Reviewer: oh I see, sgtm

    ```

   2. Boolean (True)

    ```go
    // assert boolean values
    assert.True(op.Get("settings.ipConfiguration.ipv4Enabled").Bool(), "ipv4 is enabled")
    ```

   3. GreaterOrEqual

    ```go
    // assert values that are greater than or equal to the expected value
    assert.GreaterOrEqual(op.Get("settings.dataDiskSizeGb").Float(), 10.0, "at least 5 backups are retained")
    ```

   4. Empty

    ```go
    // assert values that are supposed to be empty or nil
    assert.Empty(op.Get("settings.userLabels"), "no labels are set")
    ```

>    5. Equal

Reviewer: nit: lets add this first as this maybe most common

Author: Done.

    assert.Equal(databaseVersion, op.Get("databaseVersion").String(), "database versions is valid is set to "+databaseVersion)
    ```

The entire integration test can be found [here](https://github.com/terraform-google-modules/terraform-google-sql-db/blob/master/test/integration/mysql-public/mysql_public_test.go).

# 4. Test Execution
As mentioned in section 2 above, the blueprints test framework executes tests in two ways: auto-discovered and custom tests. Each type of test undergoes 4 stages of test execution:

1. init - runs `terraform init` and `terraform validate`
2. apply - runs `terraform apply`
3. verify - runs `terraform plan` to verify apply successful apply with no more resources to add/destroy
4. teardown - runs `terraform destroy`

By default, tests go through 4 stages above. You can also explicitly run individual stages one at a time.

## 4.1 Test prerequisites
In order for the test to execute, certain prerequisite resources and components need to be in place. These can be set up using the TF modules under `test/setup`. Running `terraform apply` in this directory will set up all resources required for the test.

```

> Note: Output values from `test/setup` are automatically loaded as Terraform environment variables and are available to both auto discovered and custom/explicit tests as inputs.

Reviewer: It might be worth cross linking this from above where we talk about setting variable values in examples.

Author: Good idea. Added.

   E.g. to run a test for just the init stage the use the following command:
   - `RUN_STAGE=init go test`

## 4.3 Auto-discovered tests
All blueprints come pre-wired with an auto-discovered test located in the `test/integration` folder. Following are the contents of the test module and can be found [here](https://github.com/terraform-google-modules/terraform-google-sql-db/blob/master/test/integration/discover_test.go) as well.

```go
package test

import (  
  // should be imported to enable testing for GO modules
  "testing"

  // should be imported to use terraform helpers in blueprints test framework
  "github.com/GoogleCloudPlatform/cloud-foundation-toolkit/infra/blueprint-test/pkg/tft"
)

// entry function for the test; can be named as Test*
func TestAll(t *testing.T) {
   // the helper to autodiscover and test blueprint examples
   tft.AutoDiscoverAndTest(t)
}
```

We’ll use the blueprint structure highlighted in section 2.1 for explaining how auto-discovered test execution works.

The auto-discovered test can be triggered as follow:
- cd into  test/integration  and run: 
  - `go test` OR
  - `go test -v`  (for verbose output)

By default, this triggers the following steps:

> - The auto-discovery module iterates through all tests defined under the `test/fixtures` directory and builds a list of tests that are only found under the `test/fixtures` directory and do not match any tests (by directory name) under `test/integration`. For this example, the following tests will be queued:

Reviewer: ```suggestion
- The auto-discovery module iterates through all tests defined under the `test/fixtures` directory and builds a list of tests that are only found under the `test/fixtures` directory and do not match any explicit tests (by directory name) under `test/integration`. For this example, the following tests will be queued:
```

Author: Changed.

   ![5](https://user-images.githubusercontent.com/21246369/131234147-b0235cc3-9f5f-4c19-892f-1b4fc6cc12f4.jpg)
   
   This illustrates the execution of the `destroy` stage where `terraform destroy` is executed to teardown all resources provisioned by the example. 
   Lastly, a status of the complete test run is shown with a tally of all passed and failed tests and eventually showing the overall status of the test run which is FAIL in this case.

## 4.4 Custom tests
Unlike auto-discovered tests, custom tests are written specifically for examples that require custom assertions and validations. Even though custom tests are run as part of the default test run as explained in section 4.2, we can also execute them in isolation as explained below. In this section we will use the integration test created in section 3.3  and show how to execute it.

1. Cd into `test/integration`
   Instead of running the whole test suite, we will target our custom test by name i.e. `TestMySqlPublicModule` in file `test/integration/mysql-public/mysql_public_test.go`
   
2. Run the one of the following commands for execution:
   - `go test -run TestMySqlPublicModule ./...`  OR

>    - `go test -run TestMySqlPublicModule ./...` -v  (for verbose output)

Reviewer: ```suggestion
   - `go test -run TestMySqlPublicModule ./... -v`   (for verbose output)
```

Author: Changed

# 1. Background and Overview

Integration testing has been instrumental in establishing a complete and functional CI pipeline. Similar to most quality software engineering modules and applications, GCP Blueprints employ an integration testing framework that is triggered as part of the CI process as well.

Apart from the necessity of including a testing framework as part of our GCP blueprints we took additional steps to ensure the following:

- Easy to learn and to setup as part of the blueprints development process
- Compatible with both Terraform and KRM based GCP blueprints

Considering the above, our test framework has been developed (details in the following sections) with backward compatibility to allow for current tests to keep functioning.

```
Note: If you have a question about the test framework, feel free to ask it on our user group. 
Feature requests can also be submitted as Issues.
```

# 2. Framework Concepts

```

> Note: The best reference documentation for the framework is the autogenerated documentation.

Author: Sure. Link added.

```

The test framework uses Golang with the testing package at its core to allow for creation and setup of GCP blueprints integration tests.

```

> Note: The test runners for the framework are published on GitHub and can be downloaded in your local workspace using `go init <package_name> & go get -t` for test development.

Reviewer: Not sure what you meant by this

Author: Actually this is not needed now. Initially I documented to explain how to setup the go module paths but we already have them setup or plan to have them setup for all repos so this is redundant. I'll take it out.

│   │   │   └── mysql_public_test.go
│   │   ├── postgresql-ha/*
│   │   ├── postgresql-public/*
│   │   ├── postgresql-public-iam/*
│   └── setup/*.tf
└── ...
```

Before jumping into the specifics for test development, let’s review how the blueprint directory is structured.
- `examples`  - this directory holds examples that may call the main blueprint or sub-blueprints (for Terraform, these are in the  modules  directory).
- `test/fixtures`  - this directory contains “fixture” configuration. In most cases, this should be configured to wrap examples that need additional inputs and unify the interface for testing. Usage of fixtures is discouraged unless necessary.
- `test/integration`  - this directory is intended to hold integration tests that are responsible for running and asserting test values for a given fixture.
- `test/setup`  - this directory holds the GCP project and initial resource setup modules that are a prerequisite for the blueprint tests to run.

# 3. Test Development
This section aims at explaining the process of developing a custom integration test which consists of the following steps:

## 3.1 Terraform based blueprints

### 3.1.1 Prepare a test example 

> The first step in the process is to create an example that leverages a TF module from the blueprint as illustrated in section 2.1. The example creation process consists of two critical steps:

Reviewer: ```suggestion
The first step in the process is to create an example that leverages a TF module from the blueprint as illustrated in section 2.1. The example creation process consists of two steps:
```

Author: Acked.

module "mysql-db" { 
 // set the source for the module being tested as part of the 
 // example
 source               = "../../modules/mysql"
 ...
}
```

4. Create a new file named `variables.tf` for organizing and defining variables that need to be passed into the example module as follows:

```hcl
...
variable "project_id" {
 description = "The ID of the project in which resources will be provisioned."
 type        = string
}
 
variable "db_name" {
 description = "The name of the SQL Database instance"
 default     = "example-mysql-public"
}
...
```

These variables are now available for access within the `main.tf` file and can be set in the example module. The following example shows how to reference the `project_id` variable in `main.tf`.

```hcl
// name for your example module
module "mysql-db" { 
 ...
 // variables required by the source module
 random_instance_name = true
 database_version     = "MYSQL_5_6"
 
 // variable being set from the example module variables configuration. 
 project_id           = var.project_id
 ...
}
```

```
Note:
Variables defined in the example module’s variables (normally variables.tf) configuration can be set 1) from 

> upstream modules calling the example module (e.g. fixtures) or 2) using environment variables by prefixing 

Reviewer: ```suggestion
wrapping modules calling the example module (e.g. fixtures) or 2) using environment variables by prefixing 
```

Author: Acked

the environment variable with “TF_VAR”. E.g. to set the project_id variable (above), setting the value in a 

> “TF_VAR_project_id” environment variable would automatically populate its value upon execution.

Reviewer: We should mention `test/setup` and how the framework exports outputs as env vars from here

Author: Added. Let me know that makes sense to you.

#### 3.1.1.2 Output variables for the test
Upon successful execution of your example module, you will most likely need outputs for resources being provisioned to validate and assert in your test. This is done using outputs in Terraform. 
1. In the `examples/mysql-public` directory create a file `outputs.tf`. The content for the file should be as follows:

```hcl
// The output value is set using the value attribute and is either computed in place or pulled from one of the recursive modules being called. In this case, it is being pulled from the mysql module that the example module is calling as its source.
...
output "mysql_user_pass" {
 value       = module.mysql-db.generated_user_password
 description = "The password for the default user. If not set, a random one will be generated and available in the generated_user_password output variable."
}
...
```

Complete code files for the example module can be found [here](https://github.com/terraform-google-modules/terraform-google-sql-db/tree/master/test/integration/mysql-public).

### 3.1.2 Write an integration test
After creating the example configuration, your example will automatically be tested and no further action is required.

> However, if you want to make custom assertions about what resources the blueprint will create, you should create an integration test in Go using the (testing)[https://pkg.go.dev/testing] package.

Reviewer: We should add a sentence stating what the custom assertions could be - Custom assertions will mostly involve making an API call to GCL (via gcloud commands) to assert a resource is configured as expected.

Author: Added.

   3. GreaterOrEqual

    ```go
    // assert values that are greater than or equal to the expected value
    assert.GreaterOrEqual(op.Get("settings.dataDiskSizeGb").Float(), 10.0, "at least 5 backups are retained")
    ```

   4. Empty

    ```go
    // assert values that are supposed to be empty or nil
    assert.Empty(op.Get("settings.userLabels"), "no labels are set")
    ```

   5. Equal

    ```go
    // assert values that are supposed to be equal to the expected values
    assert.Equal(databaseVersion, op.Get("databaseVersion").String(), "database versions is valid is set to "+databaseVersion)
    ```

The entire integration test can be found [here](https://github.com/terraform-google-modules/terraform-google-sql-db/blob/master/test/integration/mysql-public/mysql_public_test.go).

# 4. Test Execution
As mentioned in section 2 above, the blueprints test framework executes tests in two ways: auto-discovered and custom tests. Each type of test undergoes 4 stages of test execution:

1. init - runs `terraform init` and `terraform validate`
2. apply - runs `terraform apply`
3. verify - runs `terraform plan` to verify apply successful apply with no more resources to add/destroy

> 4. destroy - runs `terraform destroy`

Reviewer: ```suggestion
4. teardown - runs `terraform destroy`
```

Author: Changed.
